xtm/xtm_api.cc

```diff
@@ -0,0 +1,354 @@
+	return 0;
+}
+
+struct xtm_queue *
+xtm_queue_new(unsigned size, unsigned flags)
+{
+	int save_errno = 0;
+	struct xtm_queue *queue = (struct xtm_queue *)
+		malloc(sizeof(struct xtm_queue) +
+		       size * sizeof(union xtm_msg));
+	if (queue == NULL)
+		return NULL;
+
+	queue->flags = flags;
+	queue->is_notification_expected = false;
+
+	if (create_fds(&queue->consumer_read_fd,
+		       &queue->consumer_write_fd) < 0) {
+		save_errno = errno;
+		goto free_queue;
+	}
+
+	if ((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) != 0) {
+		if (create_fds(&queue->producer_read_fd,
+			       &queue->producer_write_fd) < 0) {
+			save_errno = errno;
+			goto close_consumer_fds;
+		}
+	} else {
+		queue->producer_read_fd = queue->producer_write_fd = -1;
+	}
+
+	if (fcntl(queue->consumer_read_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	    fcntl(queue->consumer_write_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	    ((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) != 0 &&
+	    (fcntl(queue->producer_read_fd, F_SETFL, O_NONBLOCK) < 0 ||
+	     fcntl(queue->producer_write_fd, F_SETFL, O_NONBLOCK) < 0))) {
+		save_errno = errno;
+		goto close_producer_fds;
+	}
+	if (queue->queue.create(size) < 0) {
+		save_errno = EINVAL;
+		goto close_producer_fds;
+	}
+	return queue;
+
+close_producer_fds:
+	if ((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) != 0) {
+		close(queue->producer_read_fd);
+		if (queue->producer_read_fd != queue->producer_write_fd)
+			close(queue->producer_write_fd);
+	}
+close_consumer_fds:
+	close(queue->consumer_read_fd);
+	if (queue->consumer_read_fd != queue->consumer_write_fd)
+		close(queue->consumer_write_fd);
+free_queue:
+	free(queue);
+	errno = save_errno;
+	return NULL;
+};
+
+int
+xtm_queue_delete(struct xtm_queue *queue)
+{
+	int rc = 0;
+	if (((queue->flags & XTM_QUEUE_NEED_TO_CLOSE_READFD) != 0) &&
+	    ((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) != 0) &&
+	    close(queue->producer_read_fd) < 0)
+		rc = -1;
+	if (((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) != 0) &&
+	    queue->producer_write_fd != queue->producer_read_fd &&
+	    close(queue->producer_write_fd) < 0)
+		rc = -1;
+	if (((queue->flags & XTM_QUEUE_NEED_TO_CLOSE_READFD) != 0) &&
+	    close(queue->consumer_read_fd) < 0)
+		rc = -1;
+	if (queue->consumer_read_fd != queue->consumer_write_fd &&
+	    close(queue->consumer_write_fd) < 0)
+		rc = -1;
+	free(queue);
+	return rc;
+}
+
+int
+xtm_queue_notify_consumer(struct xtm_queue *queue)
+{
+	return notify_fd(queue->consumer_write_fd);
+}
+
+int
+xtm_queue_notify_producer(struct xtm_queue *queue)
+{
+	return notify_fd(queue->producer_write_fd);
+}
+
+int
+xtm_queue_probe(struct xtm_queue *queue)
+{
+	if (queue->queue.free_count() == 0) {
+		errno = ENOBUFS;
+		return -1;
+	}
+	return 0;
+}
+
+unsigned
+xtm_queue_count(struct xtm_queue *queue)
+{
+	return queue->queue.count();
+}
+
+int
+xtm_queue_push_fun(struct xtm_queue *queue, xtm_queue_fun_t fun,
```

Comment:
Some failed calls to xtm_queue_push_fun() should not notify producer when XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS is set - HTTP requests for example (httpng would just fail them). Other requests which go into the same queue do need such notification (internal requests like "free resources" which should never fail or we are screwed). Add bool or "unsigned flags" parameter? You should probably benchmark this - HTTP request flood is likely (DDoS) and performance in such case is quite important. Additional "if" or additional useless notification?

The same for xtm_queue_push_ptr().

Reply:
Yes fixed.